### PR TITLE
Feature/8141: XHR Breakpoint Styling Changes

### LINKED ISSUE #8141: [XHR Breakpoints] Make the method selector smaller
With @Jaril's work in #8094 and https://github.com/firefox-devtools/ux/issues/45 we are going to have smaller XHR breakpoint items, 20px tall.

When going to edit mode, though, the input takes more vertical space:

1. The container's 1px border makes the height 22px instead of 20px. Maybe we should try an inset outline instead? As in: `outline: solid 1px; outline-offset: -1px;`.
2. The `<select>` itself is rather big, at least on Linux its easily more than 30px tall.

We have a minimal `<select>` style used in Firefox DevTools, for instance in the Responsive Design Mode toolbar or in the Network panel toolbar. We should reuse that style here.

I'm getting good results with this CSS:

```css
.xhr-input-method {
  margin-inline-start: 4px;
  padding-block: 0;
  padding-inline-start: 2px;
  padding-inline-end: 14px;
  border: none;
  border-radius: 2px;
  font: inherit;
  color: inherit;
  background-color: var(--toolbarbutton-background);
  -moz-appearance: none;
  appearance: none;
}
```

Then the only thing missing is adding the double-arrow icon to the right of the select.

### PR DESCRIPTION
**Pull request by blueberryapple**
Fixes #8141

### Summary of Changes

* custom styling for XHR breakpoint input field dropdown
* custom styling for `select` element, now uses nonnative styling

### Test Plan

Open debugger, focus the XHR input field; you can add the `focused` class to `.xhr-input-container` to toggle the focus manually.

- [ ] focus XHR breakpoint input field
- [ ] select an option from the dropdown
- [ ] switch themes and retest steps 1-2
- [ ] switch browsers and retests steps 1-3

#### firefox
image

image

#### chrome
image

image

---

#### firefox
image

image

#### chrome
image

image

---
### zoomed out
image

### CODE REVIEW

File: src/components/SecondaryPanes/XHRBreakpoints.css [hunk 1]
```diff
@@ -82,11 +82,36 @@
...
+  appearance: none;
+  background-color: var(--toolbarbutton-background);
+}
+
+.xhr-input-container.focused::after {
+  content: "";
+  position: relative;
+  right: 25px;
+  top: 5px;
+  width: 10px;
```
Comment (fvsch) [inline src/components/SecondaryPanes/XHRBreakpoints.css]: The original image is 8x8 pixel, and is sharp at that size. Can you use 8px here?
The mask-size can probably be `contain`.

File: src/components/SecondaryPanes/XHRBreakpoints.css [hunk 2]
```diff
@@ -82,11 +82,36 @@
...
+  -moz-appearance: none;
+  appearance: none;
+  background-color: var(--toolbarbutton-background);
+}
+
+.xhr-input-container.focused::after {
+  content: "";
+  position: relative;
+  right: 25px;
+  top: 5px;
```
Comment (fvsch) [inline src/components/SecondaryPanes/XHRBreakpoints.css]: A more robust way to center an absolutely positioned element of known height (8px) would be: `top: calc(50% - 4px)`.

File: src/components/SecondaryPanes/XHRBreakpoints.css [hunk 3]
```diff
@@ -82,11 +82,36 @@
...
+  border: none;
+  border-radius: 2px;
+  font: inherit;
+  color: inherit;
+  -moz-appearance: none;
+  appearance: none;
+  background-color: var(--toolbarbutton-background);
+}
+
+.xhr-input-container.focused::after {
```
Comment (fvsch) [inline src/components/SecondaryPanes/XHRBreakpoints.css]: I'm not fond of putting this graphical element as a pseudo-child of `.xhr-input-container` and with a somewhat arbitrary coordinate (`right: 25px`). This is likely to break at the first change. We should move this `::after` much closer to the `<select>` element, and make sure it's positioned relative to the `<select>` element and not other elements.

File: src/components/SecondaryPanes/XHRBreakpoints.css [hunk 4]
```diff
@@ -82,11 +82,36 @@
...
+  content: "";
+  position: relative;
+  right: 25px;
```
Comment (fvsch) [inline src/components/SecondaryPanes/XHRBreakpoints.css]: Can we test this UI in right-to-left direction?

We may need to add LTR/RTL styles here to handle both directions, something like:

```css
html[dir=ltr] .xhr-input-container.focused::after {
  right: 25px;
}

html[dir=rtl] .xhr-input-container.focused::after {
  left: 25px;
}
```

(Ideally we would use `inset-inline-end: 25px`, but that's not supported in Chrome yet.)

File: src/components/SecondaryPanes/XHRBreakpoints.css (hunk 1 above)
Comment (blueberryapple) [reply src/components/SecondaryPanes/XHRBreakpoints.css]: Okay, fixed!

File: src/components/SecondaryPanes/XHRBreakpoints.css (hunk 2 above)
Comment (blueberryapple) [reply src/components/SecondaryPanes/XHRBreakpoints.css]: Alrighty, done!

File: src/components/SecondaryPanes/XHRBreakpoints.css (hunk 3 above)
Comment (blueberryapple) [reply src/components/SecondaryPanes/XHRBreakpoints.css]: I made a wrapper span for the select element.

File: src/components/SecondaryPanes/XHRBreakpoints.css (hunk 4 above)
Comment (blueberryapple) [reply src/components/SecondaryPanes/XHRBreakpoints.css]: image

I put the rules in and tested for RTL layout.

**Comment by blueberryapple** [comment]
I'm getting a failing test for this test:
```
  ● XHR Breakpoints › should submit the URL and method when adding a breakpoint

    expect(jest.fn()).toHaveBeenCalledWith(expected)

    Expected mock function to have been called with:
      "ANY"
    as argument 2, but it was called with
      undefined.

    Difference:

      Comparing two different types of values. Expected string but received undefined.

      284 | 
      285 |       xhrBreakpointsComponent.find("form").simulate("submit", mockEvent);
    > 286 |       expect(setXHRBreakpointCallback).toHaveBeenCalledWith(urlValue, method);
          |                                        ^
      287 |     });
      288 |   });
      289 | 

      at src/components/SecondaryPanes/tests/XHRBreakpoints.spec.js:286:40
          at Array.forEach (<anonymous>)
      at Object.<anonymous> (src/components/SecondaryPanes/tests/XHRBreakpoints.spec.js:270:25)
```

I've added the wrapper element, and am pretty sure that that's what's causing this to fail :cry:, so I've tried debugging this with the jest vscode plugin, but get a lot of these errors:
```
Starting inspector on 127.0.0.1:14824 failed: address already in use
```

and my breakpoints then never hit.

**Comment by fvsch** [comment]
It looks like the DOM structure is somewhat hardcoded here:
https://github.com/firefox-devtools/debugger/blob/4c6f1573cecd14fb983ef55d3dc19d51dab7457f/src/components/SecondaryPanes/XHRBreakpoints.js#L116

Changing that line to `{ inputMethod: e.target.querySelector(".xhr-input-method").value }` seems to fix the issue.

@jaril, do you know if there is a cleaner fix we should apply?

**Comment by darkwing** [comment]
Great work @blueberryapple and thank you for jump in to advise @fvsch!  I look forward to seeing where this goes! 👍

**Comment by blueberryapple** [comment]
Okay, I resolved the failing tests on circleci with help from @fvsch, but some mochitests are failing, they all fail due to `helpers.js 1308:16` for each of the failing tests. I'm suspicious that it's also the selector again, but am not sure where they are defined. Here is one of the failing tests' output:
```
 TEST-UNEXPECTED-FAIL helpers.js:1308 
 TypeError: popup is null
     selectContextMenuItem browser/devtools/client/debugger/new/test/mochitest/helpers.js 1308:16
      browser/devtools/client/debugger/new/test/mochitest/browser_dbg-breakpoints-actions.js 19:3
     Async*Tester_execTest/< browser-test.js 1106:34
     Tester_execTest browser-test.js 1134:12
     nextTest/< browser-test.js 995:14
     SimpleTest.waitForFocus/waitForFocusInner/focusedOrLoaded/< tests/SimpleTest/SimpleTest.js 803:59
```

**Comment by darkwing** [comment]
Nice work @blueberryapple !  I'll have a look today! :)

**Comment by darkwing** [comment]
Could you rebase this @blueberryapple?  Thank you!

**Comment by blueberryapple** [comment]
Okay, I've rebased on top of master, but I'm getting these failing tests:
```
Summary of all failing tests
 FAIL  src/utils/sources-tree/tests/sortTree.spec.js
  ● sources-tree › sortEntireTree › alphabetically sorts children

    TypeError: (0 , _index.sortEntireTree) is not a function

      32 |       addToTree(_tree, source2, "http://example.com/", "");
      33 |       addToTree(_tree, source3, "http://example.com/", "");
    > 34 |       const tree = sortEntireTree(_tree);
         |                    ^
      35 | 
      36 |       const base = tree.contents[0];
      37 |       const fooNode = base.contents[0];

      at Object.it (src/utils/sources-tree/tests/sortTree.spec.js:34:20)

  ● sources-tree › sortEntireTree › sorts folders first

    TypeError: (0 , _index.sortEntireTree) is not a function

      64 |         addToTree(_tree, source, "http://example.com/", "")
      65 |       );
    > 66 |       const tree = sortEntireTree(_tree);
         |                    ^
      67 |       const domain = tree.contents[0];
      68 | 
      69 |       const [

      at Object.it (src/utils/sources-tree/tests/sortTree.spec.js:66:20)

  ● sources-tree › sortEntireTree › puts folder at the top of the sort

    TypeError: (0 , _index.sortEntireTree) is not a function

      105 |         addToTree(_tree, source, "http://example.com/", "")
      106 |       );
    > 107 |       const tree = sortEntireTree(_tree);
          |                    ^
      108 |       const [
      109 |         bFolderNode,
      110 |         cFolderNode,

      at Object.it (src/utils/sources-tree/tests/sortTree.spec.js:107:20)

  ● sources-tree › sortEntireTree › puts root debugee url at the top of the sort

    TypeError: (0 , _index.sortEntireTree) is not a function

      139 |         addToTree(_treeB, source, rootB, "");
      140 |       });
    > 141 |       const treeA = sortEntireTree(_treeA, rootA);
          |                     ^
      142 |       const treeB = sortEntireTree(_treeB, rootB);
      143 | 
      144 |       expect(treeA.contents[0].contents[0].name).toBe("b.js");

      at Object.it (src/utils/sources-tree/tests/sortTree.spec.js:141:21)

 FAIL  src/utils/tests/empty-lines.spec.js
  ● emptyLines › no positions

    expect(received).toEqual(expected)

    Expected value to equal:
      [1, 2]
    Received:
      [1]

    Difference:

    - Expected
    + Received

      Array [
        1,
    -   2,
      ]

      17 |     const source = makeSource("foo", { text: "\n" });
      18 |     const lines = findEmptyLines(source, []);
    > 19 |     expect(lines).toEqual([1, 2]);
         |                   ^
      20 |   });
      21 | 
      22 |   it("one position", () => {

      at Object.it (src/utils/tests/empty-lines.spec.js:19:19)

  ● emptyLines › one position

    TypeError: (0 , _sourceMaps.getSelectedLocation) is not a function

      22 | 
      23 |   const breakpointLines = breakpointPositions
    > 24 |     .map(point => getSelectedLocation(point, source).line)
         |                   ^
      25 |     // NOTE: at the moment it is possible the location is an unmapped generated
      26 |     // line which could be greater than the line count.
      27 |     .filter(line => line <= lineCount);

      at breakpointPositions.map.point (src/utils/empty-lines.js:24:19)
          at Array.map (<anonymous>)
      at findEmptyLines (src/utils/empty-lines.js:24:6)
      at Object.it (src/utils/tests/empty-lines.spec.js:24:19)

  ● emptyLines › outside positions are not included

    TypeError: (0 , _sourceMaps.getSelectedLocation) is not a function

      22 | 
      23 |   const breakpointLines = breakpointPositions
    > 24 |     .map(point => getSelectedLocation(point, source).line)
         |                   ^
      25 |     // NOTE: at the moment it is possible the location is an unmapped generated
      26 |     // line which could be greater than the line count.
      27 |     .filter(line => line <= lineCount);

      at breakpointPositions.map.point (src/utils/empty-lines.js:24:19)
          at Array.map (<anonymous>)
      at findEmptyLines (src/utils/empty-lines.js:24:6)
      at Object.it (src/utils/tests/empty-lines.spec.js:30:19)

 FAIL  src/actions/pause/tests/pause.spec.js
  ● pause › stepping › should step over when paused

    Cannot spy the getNextStep property because it is not a function; undefined given instead

      148 |       await dispatch(actions.paused(mockPauseInfo));
      149 |       const cx = selectors.getThreadContext(getState());
    > 150 |       const getNextStepSpy = jest.spyOn(parser, "getNextStep");
          |                                   ^
      151 |       dispatch(actions.stepOver(cx));
      152 |       expect(getNextStepSpy).not.toBeCalled();
      153 |       expect(isStepping(getState(), "FakeThread")).toBeTruthy();

      at ModuleMockerClass.spyOn (node_modules/jest-mock/build/index.js:697:15)
      at Object.<anonymous> (src/actions/pause/tests/pause.spec.js:150:35)
      at step (src/actions/pause/tests/pause.spec.js:13:191)
      at src/actions/pause/tests/pause.spec.js:13:361

  ● pause › stepping › should step over when paused before an await

    Cannot spy the getNextStep property because it is not a function; undefined given instead

      167 |       await dispatch(actions.paused(mockPauseInfo));
      168 |       const cx = selectors.getThreadContext(getState());
    > 169 |       const getNextStepSpy = jest.spyOn(parser, "getNextStep");
          |                                   ^
      170 |       dispatch(actions.stepOver(cx));
      171 |       expect(getNextStepSpy).toBeCalled();
      172 |       getNextStepSpy.mockRestore();

      at ModuleMockerClass.spyOn (node_modules/jest-mock/build/index.js:697:15)
      at Object.<anonymous> (src/actions/pause/tests/pause.spec.js:169:35)
      at step (src/actions/pause/tests/pause.spec.js:13:191)
      at src/actions/pause/tests/pause.spec.js:13:361

  ● pause › stepping › should step over when paused after an await

    Cannot spy the getNextStep property because it is not a function; undefined given instead

      189 |       await dispatch(actions.paused(mockPauseInfo));
      190 |       const cx = selectors.getThreadContext(getState());
    > 191 |       const getNextStepSpy = jest.spyOn(parser, "getNextStep");
          |                                   ^
      192 |       dispatch(actions.stepOver(cx));
      193 |       expect(getNextStepSpy).toBeCalled();
      194 |       getNextStepSpy.mockRestore();

      at ModuleMockerClass.spyOn (node_modules/jest-mock/build/index.js:697:15)
      at Object.<anonymous> (src/actions/pause/tests/pause.spec.js:191:35)
      at step (src/actions/pause/tests/pause.spec.js:13:191)
      at src/actions/pause/tests/pause.spec.js:13:361

  ● pause › resumed › resuming - will re-evaluate watch expressions

    TypeError: Cannot read property 'hasSyntaxError' of undefined

      40 |     }
      41 | 
    > 42 |     const expressionError = await evaluationsParser.hasSyntaxError(input);
         |                                                     ^
      43 | 
      44 |     const expression = getExpression(getState(), input);
      45 |     if (expression) {

      at src/actions/expressions.js:42:53
      at step (src/actions/expressions.js:25:191)
      at src/actions/expressions.js:25:437
      at src/actions/expressions.js:25:99
      at src/actions/expressions.js:37:3
      at action (src/actions/utils/middleware/thunk.js:24:11)
      at Object.<anonymous> (src/actions/pause/tests/pause.spec.js:397:13)
      at step (src/actions/pause/tests/pause.spec.js:13:191)
      at src/actions/pause/tests/pause.spec.js:13:361

 FAIL  src/actions/tests/expressions.spec.js (7.431s)
  ● expressions › should add an expression

    TypeError: Cannot read property 'hasSyntaxError' of undefined

      40 |     }
      41 | 
    > 42 |     const expressionError = await evaluationsParser.hasSyntaxError(input);
         |                                                     ^
      43 | 
      44 |     const expression = getExpression(getState(), input);
      45 |     if (expression) {

      at src/actions/expressions.js:42:53
      at step (src/actions/expressions.js:25:191)
      at src/actions/expressions.js:25:437
      at src/actions/expressions.js:25:99
      at src/actions/expressions.js:37:3
      at action (src/actions/utils/middleware/thunk.js:24:11)
      at Object.<anonymous> (src/actions/tests/expressions.spec.js:57:11)
      at step (src/actions/tests/expressions.spec.js:28:191)
      at src/actions/tests/expressions.spec.js:28:437
      at Object.<anonymous> (src/actions/tests/expressions.spec.js:28:99)

  ● expressions › should not add invalid expressions

    TypeError: Cannot read property 'hasSyntaxError' of undefined

      40 |     }
      41 | 
    > 42 |     const expressionError = await evaluationsParser.hasSyntaxError(input);
         |                                                     ^
      43 | 
      44 |     const expression = getExpression(getState(), input);
      45 |     if (expression) {

      at src/actions/expressions.js:42:53
      at step (src/actions/expressions.js:25:191)
      at src/actions/expressions.js:25:437
      at src/actions/expressions.js:25:99
      at src/actions/expressions.js:37:3
      at action (src/actions/utils/middleware/thunk.js:24:11)
      at Object.<anonymous> (src/actions/tests/expressions.spec.js:71:11)
      at step (src/actions/tests/expressions.spec.js:28:191)
      at src/actions/tests/expressions.spec.js:28:437
      at Object.<anonymous> (src/actions/tests/expressions.spec.js:28:99)

  ● expressions › should update an expression

    TypeError: Cannot read property 'hasSyntaxError' of undefined

      40 |     }
      41 | 
    > 42 |     const expressionError = await evaluationsParser.hasSyntaxError(input);
         |                                                     ^
      43 | 
      44 |     const expression = getExpression(getState(), input);
      45 |     if (expression) {

      at src/actions/expressions.js:42:53
      at step (src/actions/expressions.js:25:191)
      at src/actions/expressions.js:25:437
      at src/actions/expressions.js:25:99
      at src/actions/expressions.js:37:3
      at action (src/actions/utils/middleware/thunk.js:24:11)
      at Object.<anonymous> (src/actions/tests/expressions.spec.js:80:11)
      at step (src/actions/tests/expressions.spec.js:28:191)
      at src/actions/tests/expressions.spec.js:28:437
      at Object.<anonymous> (src/actions/tests/expressions.spec.js:28:99)

  ● expressions › should not update an expression w/ invalid code

    TypeError: Cannot read property 'hasSyntaxError' of undefined

      40 |     }
      41 | 
    > 42 |     const expressionError = await evaluationsParser.hasSyntaxError(input);
         |                                                     ^
      43 | 
      44 |     const expression = getExpression(getState(), input);
      45 |     if (expression) {

      at src/actions/expressions.js:42:53
      at step (src/actions/expressions.js:25:191)
      at src/actions/expressions.js:25:437
      at src/actions/expressions.js:25:99
      at src/actions/expressions.js:37:3
      at action (src/actions/utils/middleware/thunk.js:24:11)
      at Object.<anonymous> (src/actions/tests/expressions.spec.js:90:11)
      at step (src/actions/tests/expressions.spec.js:28:191)
      at src/actions/tests/expressions.spec.js:28:437
      at Object.<anonymous> (src/actions/tests/expressions.spec.js:28:99)

  ● expressions › should delete an expression

    TypeError: Cannot read property 'hasSyntaxError' of undefined

      40 |     }
      41 | 
    > 42 |     const expressionError = await evaluationsParser.hasSyntaxError(input);
         |                                                     ^
      43 | 
      44 |     const expression = getExpression(getState(), input);
      45 |     if (expression) {

      at src/actions/expressions.js:42:53
      at step (src/actions/expressions.js:25:191)
      at src/actions/expressions.js:25:437
      at src/actions/expressions.js:25:99
      at src/actions/expressions.js:37:3
      at action (src/actions/utils/middleware/thunk.js:24:11)
      at Object.<anonymous> (src/actions/tests/expressions.spec.js:99:11)
      at step (src/actions/tests/expressions.spec.js:28:191)
      at src/actions/tests/expressions.spec.js:28:437
      at Object.<anonymous> (src/actions/tests/expressions.spec.js:28:99)

  ● expressions › should evaluate expressions global scope

    TypeError: Cannot read property 'hasSyntaxError' of undefined

      40 |     }
      41 | 
    > 42 |     const expressionError = await evaluationsParser.hasSyntaxError(input);
         |                                                     ^
      43 | 
      44 |     const expression = getExpression(getState(), input);
      45 |     if (expression) {

      at src/actions/expressions.js:42:53
      at step (src/actions/expressions.js:25:191)
      at src/actions/expressions.js:25:437
      at src/actions/expressions.js:25:99
      at src/actions/expressions.js:37:3
      at action (src/actions/utils/middleware/thunk.js:24:11)
      at Object.<anonymous> (src/actions/tests/expressions.spec.js:111:11)
      at step (src/actions/tests/expressions.spec.js:28:191)
      at src/actions/tests/expressions.spec.js:28:437
      at Object.<anonymous> (src/actions/tests/expressions.spec.js:28:99)

  ● expressions › should evaluate expressions in specific scope

    TypeError: Cannot read property 'hasSyntaxError' of undefined

      40 |     }
      41 | 
    > 42 |     const expressionError = await evaluationsParser.hasSyntaxError(input);
         |                                                     ^
      43 | 
      44 |     const expression = getExpression(getState(), input);
      45 |     if (expression) {

      at src/actions/expressions.js:42:53
      at step (src/actions/expressions.js:25:191)
      at src/actions/expressions.js:25:437
      at src/actions/expressions.js:25:99
      at src/actions/expressions.js:37:3
      at action (src/actions/utils/middleware/thunk.js:24:11)
      at Object.<anonymous> (src/actions/tests/expressions.spec.js:127:11)
      at step (src/actions/tests/expressions.spec.js:28:191)
      at src/actions/tests/expressions.spec.js:28:361
```

**Comment by fvsch** [comment]
@darkwing Do you know how we could move forward on this?
Should we move this issue to Bugzilla and the patch to Phabricator, if possible?

**Comment by darkwing** [comment]
I've moved this to https://phabricator.services.mozilla.com/D38692 ; please check out there !  Thank you!

**Pull request closed without merging**